### PR TITLE
Fixed text indexing issue for :class:`~.Text`

### PR DESCRIPTION
**Pull request by xia0long**
## Overview: What does this pull request change?

+ Rewrite function find_indexes, removed the feature of using indices in `word`.
+ Add an offset to indexes when `self.disable_ligatures == False`.
+ Add test case for Text.set_color_by_t2c().

## Motivation and Explanation: Why and how do your changes improve the library?

1. Get the correct indexes of words that need to be colored.

    code used:
    
    ```py
    class Test(Scene):
        def construct(self):
            t = Text("Cherry Blossoms in Spring", 
                disable_ligatures=False, 
                t2c={"Cherry": RED, "Spring": GREEN}
            ).scale(1)
    
            self.add(t)
    ```
  
    before:
    before
    
    after:
    after

2. Removed the feature of using indices in `word` in function `find_indexes`,  putting slices in this position is odd and can be problematic in some cases.
    
    code used:
    
    ```py
    class Test(Scene):
        def construct(self):
            t = Text("hello [:2]", 
                disable_ligatures=False, 
                t2c={"[:2]": YELLOW, "o": RED}
            ).scale(1)
    
            self.add(t)
    ```
    image

    There is a conflict between the two approaches.

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by ryanmccauley211** [comment]
Thanks for raising this, can you add a code snippet to replicate the error and maybe a bit more detail on what the indexing error is for anyone who is unfamiliar with it please?

**Comment by xia0long** [comment]
> Thanks for raising this, can you add a code snippet to replicate the error and maybe a bit more detail on what the indexing error is for anyone who is unfamiliar with it please?

Hi @ryanmccauley211 !
I added some descriptions, feel free to contact me if u have any further infomation.

**Review by ryanmccauley211 — CHANGES_REQUESTED** [review]
It works well for me with the example you give and fixes the index issue, but it seems to ignore coloring when spaces are added. For example:

```py
class Test(Scene):
    def construct(self):
        t = Text("Cherry Blossoms in Spring",
            disable_ligatures=False,
            t2c={"Cherry Blossoms": RED, "Spring": GREEN}
        ).scale(1)

        self.add(t)
```

currently results in the broken indexing but adds color to both words:

Screenshot 2021-08-30 at 11 13 02

With your change it doesn't seem to add any color:

Screenshot 2021-08-30 at 11 16 26

**Comment by WampyCakes** [comment]
Since it seems that our text indexing has had numerous problems, would you mind adding a test to ensure we catch this should any future problems arise?

**Comment by hydrobeam** [comment]
I'm not actually sure we can properly test `Text` right now since the default font it uses varies based on operating system. Maybe you can get around that by specifying a font shared across the OS's we test?

**Comment by xia0long** [comment]
@WampyCakes 

No problem, I'll add a basic test which we can improve in the future.

**Comment by xia0long** [comment]
When I wrote the test case, It seems that the `color` attribute in `Text` can't get the correct value, so I use `get_color()` instead.

```py
In [3]: text = Text("Hello World", t2c={"Hello": RED})

In [4]: text.color
Out[4]: <Color white>

In [5]: text.submobjects[1].color
Out[5]: <Color white>

In [6]: text.submobjects[1].get_color()
Out[6]: <Color #fc6255>
```

**Review by xia0long — COMMENTED** [review]
fixed

**Review by ryanmccauley211 — COMMENTED** [review]
It is working really well now thank you! It seems to have failed the build though, once that is resolved I think it looks good to approve

File: manim/mobject/svg/text_mobject.py [hunk 1]
```diff
@@ -572,19 +572,7 @@ def gen_chars(self):
...
-            start = len(text) + start if start < 0 else start
-            end = len(text) + end if end < 0 else end
-            return [(start, end)]
-        indexes = []
-        index = text.find(word)
-        while index != -1:
-            indexes.append((index, index + len(word)))
-            index = text.find(word, index + len(word))
-        return indexes
+        return [[i.start(), i.start() + len(word)] for i in re.finditer(word, text)]
```
[the remark this answers is not in the document]
Comment (xia0long) [reply manim/mobject/svg/text_mobject.py]: Hi @naveen521kk! Thank you for your feedback, can you give an example of this problem?

[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/mobject/svg/text_mobject.py]: Something like this
```py
class a(Scene):
  def construct(self):
    self.add(Text('hello', t2c={'[2:4]':RED}))
```
should work.

[the remark this answers is not in the document]
Comment (xia0long) [reply manim/mobject/svg/text_mobject.py]: This approach does help with long string coloring, but putting slices in this position is odd and can be problematic in some cases.

**Code**
```py
class Test(Scene):
    def construct(self):
        t = Text("hello [:2]", 
            disable_ligatures=False, 
            t2c={"[:2]": YELLOW, "o": RED}
        ).scale(1)

        self.add(t)
```
Test_ManimCE_v0 10 0

[the remark this answers is not in the document]
Comment (xia0long) [reply manim/mobject/svg/text_mobject.py]: When a long string needs to be colored, I think the approach blow is more appropriate.
```py
class Test(Scene):
    def construct(self):
        text_string = "Long string"
        self.add(Text(text_string, t2c={text_string[:3]:RED}))
```

**Comment by xia0long** [comment]
> It works well for me with the example you give and fixes the index issue, but it seems to ignore coloring when spaces are added. For example:
> 
> ```python
> class Test(Scene):
>     def construct(self):
>         t = Text("Cherry Blossoms in Spring",
>             disable_ligatures=False,
>             t2c={"Cherry Blossoms": RED, "Spring": GREEN}
>         ).scale(1)
> 
>         self.add(t)
> ```
> 
> currently results in the broken indexing but adds color to both words:
> 
> Screenshot 2021-08-30 at 11 13 02
> 
> With your change it doesn't seem to add any color:
> 
> Screenshot 2021-08-30 at 11 16 26

Hi @ryanmccauley211, thanks for your review! and these requested changes are outdated.

File: manim/mobject/svg/text_mobject.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (naveen521kk) [reply manim/mobject/svg/text_mobject.py]: You would need to document them and also in changelog.

[the remark this answers is not in the document]
Comment (xia0long) [reply manim/mobject/svg/text_mobject.py]: Sorry for the delayed response! I have made some changes,  feel free to contact me if you have any further information.

**Comment by ryanmccauley211** [comment]
@xia0long @naveen521kk Is there anything still outstanding in this PR? The docs are still failing though so that needs fixed before we can merge it.

**Comment by naveen521kk** [comment]
This PR removes support for indexes in `t2c` as described https://github.com/ManimCommunity/manim/pull/1977#discussion_r700460673. I don't like to break that feature (or atleast make a clear discussion with others before breaking it).

**Comment by xia0long** [comment]
The reason for removing support for indexes I've written [here](https://github.com/ManimCommunity/manim/pull/1977#discussion_r700769448), perhaps we can discuss whether there is a better way to fix this problem.

**Comment by ryanmccauley211** [comment]
> The reason for removing support for indexes I've written [here](https://github.com/ManimCommunity/manim/pull/1977#discussion_r700769448), perhaps we can discuss whether there is a better way to fix this problem.

Ah sorry I didn't realise the indexing part was still an issue. Yea I agree with naveen that we should try to find a fix that doesn't remove the option to use indexes

**Pull request closed without merging**